### ISSUE TITLE
is It expected to sent App Hangs event if System popup about FaceId permission shown ?

### ISSUE THREAD

**Issue by expcapitaldev** [getsentry/sentry-cocoa#3472]
### Platform

iOS

### Environment

Develop

### Installed

Carthage

### Version

8.13.1

### Steps to Reproduce

Show System Popup with Access to Face Id  after `Secret.load`

```Swift
func loadSecret(_ command: CDVInvokedUrlCommand) {
    let data  = command.arguments[0] as AnyObject?;
    var prompt = "Authentication"
    if let description = data?.object(forKey: "description") as! String? {
        prompt = description;
    }
    var pluginResult: CDVPluginResult
    do {
        **let result = try Secret().load(prompt)**
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: result);
    } catch {
        var code = PluginError.BIOMETRIC_UNKNOWN_ERROR.rawValue
        var message = error.localizedDescription
        if let err = error as? KeychainError {
            code = err.pluginError.rawValue
            message = err.localizedDescription
        }
        let errorResult = ["code": code, "message": message] as [String : Any]
        // ...
    }
}
```

### Expected Result

no ANR App Hangs event

### Actual Result

App Hanging
 
Secret.load

**Comment by brustolin** [comment]
Hello @expcapitaldev thanks for reaching out. 
I agree that this is not the desired behaviour, we need to investigate whether its possible to detect this system events to avoid reporting app hangs.

**Comment by kahest** [comment]
The system notifications for iPhone Pro with dynamic island are not causing the app to hang according to our preliminary tests. We'll test other models without dynamic island next.

**Comment by expcapitaldev** [comment]
hi guys,  any updates here ?

**Comment by brustolin** [comment]
Unfortunately we could not reproduce this yet @expcapitaldev 
If you have any suggestion to help us out we appreciate.

**Comment by expcapitaldev** [comment]
Show System Popup with Access to Face Id after Secret.load
So, you need only open that System Popup about access to FaceId, I will try later send screenshots, but please show it and see

**Comment by julianD77** [comment]
We noticed our App started receiving instances of these errors for our live users after we recently updated the `@sentry/react-native` plugin from `4.15.0` to `5.9.1`. 

When looking into the cause we then found this issue, alongside some similar posts: 

https://github.com/getsentry/sentry-cocoa/discussions/2715
https://github.com/react-native-clipboard/clipboard/issues/212
https://github.com/flutter/flutter/issues/133557#issuecomment-1870403283

We have been able to confirm that calls to `Clipboard.getString` and `Clipboard.getStrings` are resulting in "App hanging for at least ..." errors being reported to Sentry.

We have confirmed this using a specific test build which makes calls to the Clipboard method(s) `Clipboard.getString(s)`. These methods can cause (see below) the iOS `UIPasteboard` permissions prompt to appear and will cause the `SentryANRTracker` to report an ANR and an "App hanging" error to be reported in Sentry.

NOTE that in order for the iOS System Privacy prompt to appear you may need to do the following:

- Settings/General/Transfer & Reset/Reset/Reset Location & Privacy
- Reboot the device
- Delete & Reinstall the iOS test App being used to trigger the privacy prompt
- Use a Release build of the test App

**Comment by philipphofmann** [comment]
Thanks for the detailed info, @julianD77.

**Comment by philipphofmann** [comment]
I tried to reproduce the issue but wasn't able to. I tried push notification permission, location permission, and a biometric prompt in our iOS-Swift sample app, see https://github.com/getsentry/sentry-cocoa/blob/27039af9ece2efea572500caac518ac177f85cba/Samples/iOS-Swift/iOS-Swift/ViewControllers/PermissionsViewController.swift#L83-L110

I also tried all of these with a TestFlight build, but none of these triggered an ANR. 

@SrAdam, for which type of app do you experience the issue? A native iOS app? If yes, does it use SwiftUI or UIViewControllers? Do you build your app with React-Native or Flutter?

@expcapitaldev, when looking at your code snippet you posted above, I assume you use Cordova? Is that correct? If yes, doest this only happen with Cordova, or does it also happen when interacting with native iOS code?

**Comment by SrAdam** [comment]
Thank you Philipp for the quick answer, and let me say sorry for not being explicit enough.

I am running a flutter app. My app was last updated 20 days ago, accordingly my versions:

Pubspec.lock:
```
dependencies:
  sentry_flutter: ^7.14.0
sdks:
  dart: ">=3.2.0 <4.0.0"
  flutter: ">=3.16.0"
```

Podfile.lock:
```
  - Sentry/HybridSDK (8.17.2):
    - SentryPrivate (= 8.17.2)
  - sentry_flutter (0.0.1):
    - Flutter
    - FlutterMacOS
    - Sentry/HybridSDK (= 8.17.2)
  - SentryPrivate (8.17.2)
```

**Comment by expcapitaldev** [comment]
Thank you Philipp for answer, now I very very busy  and will try to answer later for all questions and provide more code and information about case

**Comment by caspg** [comment]
How do you deal with this in the meantime? Setting `enableAppHangTracking` to false?

**Comment by kjxbyz** [comment]
same issues

<img width="1169" alt="2024-05-19 07 08 41" src="https://github.com/getsentry/sentry-cocoa/assets/47768002/d3742179-47f9-48f1-9642-cb2a21c8744c">

<img width="1179" alt="2024-05-19 07 08 52" src="https://github.com/getsentry/sentry-cocoa/assets/47768002/70cb27ce-54ad-4e8b-85a7-725b1299128f">

**Comment by philipphofmann** [comment]
Thanks for the update, @kjxbyz.

**Comment by philipphofmann** [comment]
I can reproduce the problem by simply calling in our iOS-Swift sample app.
```
if let clipboard = UIPasteboard.general.string {

}
```

When the following dialog appears, the Cocoa SDK wrongly detects an app hang. So this has nothing to do with React-Native or Flutter.
![CleanShot 2024-06-13 at 15 01 49@2x](https://github.com/getsentry/sentry-cocoa/assets/2443292/dc2592c9-2c23-427e-b9e4-ef61a7dbd2f4)

@brustolin, I tried to figure out how we can detect such dialogs. Watching the `willResignActiveNotification` and `didBecomeActiveNotificationName` doesn't work. I also played around with the FramesTracker to check if the CADisplayLink's properties timestamp, targetTimestamp, duration, or paused give away any useful information to know if a permission dialog is opened, but sadly, these properties all behave exactly the same as when I press the ANR fully blocking button. Do you maybe have any ideas?

**Comment by brustolin** [comment]
Good thing we can reproduce it now. 

Unfortunately I dont think there is an automatic solution for this, since Apple doesn't provide any API to report a system dialog being presented.

Some alternatives:
- Swizzle all potencial blocking functions (Clipboard, TouchId/FaceId, Location)
- Provide an automatic API that should be called before using one of those system functions.


 ```swift
 SentrySDK.pauseAppHangTracking {
 	//call clipboard in here
 }
 ```
  or
 ```swift
 SentrySDK.pauseAppHangTracking()
 //call clipboard in here
 SentrySDK.resumeAppHangTracking()
 ```
 Not a friendly solution but this is Apples fault, nothing we can do about it.

**Comment by philipphofmann** [comment]
We could start with the `pauseAppHangTracking` callbacks for now, but they are also suboptimal cause most people have to learn it the hard way they have to use them. I don't really want to swizzle the blocking functions.

**Comment by philipphofmann** [comment]
In the upcoming release, you can use the following API to ignore the app hangs. We're going to add these APIs also to Flutter and RN.

```swift
SentrySDK.pauseAppHangTracking()
// Do something that might cause the app to hang,
// and you don't want the Cocoa SDK to report it.
SentrySDK.resumeAppHangTracking()
```

**Comment by expcapitaldev** [comment]
thanks!

**Comment by philipphofmann** [comment]
Meta team SDK issue for reference: https://github.com/getsentry/team-mobile/issues/186.

**Comment by antonengelhardt** [comment]
@philipphofmann Could you please comment on this issue when its implemented in flutter? On version 9.6.0, i cannot use this method.

**Comment by kahest** [comment]
@antonengelhardt this was released for Flutter with SDK version [8.4.0](https://github.com/getsentry/sentry-dart/releases/tag/8.4.0) - you should be able to use `SentryFlutter.pauseAppHangTracking()` and `SentryFlutter.resumeAppHangTracking()`. If this isn't working for you, please [open an issue on sentry-dart](https://github.com/getsentry/sentry-dart/issues/new/choose).

**Comment by antonengelhardt** [comment]
Thanks! I used `Sentry` and not `SentryFlutter`, that fixed it
